src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,539 @@
+    exact compatible G hc hc'
+  end,
+  ..homeomorph.refl M }
+
+/-- The inverse of a diffeomorphism is a diffeomorphism -/
+def diffeomorph.symm (e : diffeomorph G M M') : diffeomorph G M' M :=
+{ smooth_to_fun := begin
+    assume c c' hc hc',
+    have : (c'.symm ≫ₕ e.to_homeomorph.to_local_homeomorph ≫ₕ c).symm ∈ G :=
+      G.inv _ (e.smooth_to_fun c' c hc' hc),
+    simp at this,
+    rwa [trans_symm_eq_symm_trans_symm, trans_symm_eq_symm_trans_symm, symm_symm, trans_assoc]
+      at this,
+  end,
+  ..e.to_homeomorph.symm}
+
+/-- The composition of diffeomorphisms is a diffeomorphism -/
+def diffeomorph.trans (e : diffeomorph G M M') (e' : diffeomorph G M' M'') : diffeomorph G M M'' :=
+{ smooth_to_fun := begin
+    /- Let c and c' be two charts in M and M''. We want to show that e' ∘ e is smooth in these
+    charts, around any point x. For this, let y = e (c⁻¹ x), and consider a chart g around y.
+    Then g ∘ e ∘ c⁻¹ and c' ∘ e' ∘ g⁻¹ are both smooth as e and e' are diffeomorphisms, so
+    their composition is smooth, and it coincides with c' ∘ e' ∘ e ∘ c⁻¹ around x. -/
+    assume c c' hc hc',
+    refine G.locality _ (λx hx, _),
+    let f₁ := e.to_homeomorph.to_local_homeomorph,
+    let f₂ := e'.to_homeomorph.to_local_homeomorph,
+    let f  := (e.to_homeomorph.trans e'.to_homeomorph).to_local_homeomorph,
+    have feq : f = f₁ ≫ₕ f₂ := homeomorph.trans_to_local_homeomorph _ _,
+    -- define the atlas g around y
+    let y := (c.symm ≫ₕ f₁).to_fun x,
+    let g := chart_at H y,
+    have hg₁ := chart_mem_atlas H y,
+    have hg₂ := mem_chart_source H y,
+    let s := (c.symm ≫ₕ f₁).source ∩ (c.symm ≫ₕ f₁).to_fun ⁻¹' g.source,
+    have open_s : is_open s,
+      by apply (c.symm ≫ₕ f₁).continuous_to_fun.preimage_open_of_open; apply open_source,
+    have : x ∈ s,
+    { split,
+      { simp only [trans_source, preimage_univ, inter_univ, homeomorph.to_local_homeomorph_source],
+        rw trans_source at hx,
+        exact hx.1 },
+      { exact hg₂ } },
+    refine ⟨s, open_s, ⟨this, _⟩⟩,
+    let F₁ := (c.symm ≫ₕ f₁ ≫ₕ g) ≫ₕ (g.symm ≫ₕ f₂ ≫ₕ c'),
+    have A : F₁ ∈ G := G.comp _ _ (e.smooth_to_fun c g hc hg₁) (e'.smooth_to_fun g c' hg₁ hc'),
+    let F₂ := (c.symm ≫ₕ f ≫ₕ c').restr s,
+    have : F₁ ≈ F₂ := calc
+      F₁ ≈ c.symm ≫ₕ f₁ ≫ₕ (g ≫ₕ g.symm) ≫ₕ f₂ ≫ₕ c' : by simp [F₁, trans_assoc]
+      ... ≈ c.symm ≫ₕ f₁ ≫ₕ (of_set g.source g.open_source) ≫ₕ f₂ ≫ₕ c' :
+        by simp [eq_on_source_trans, trans_self_symm g]
+      ... ≈ ((c.symm ≫ₕ f₁) ≫ₕ (of_set g.source g.open_source)) ≫ₕ (f₂ ≫ₕ c') :
+        by simp [trans_assoc]
+      ... ≈ ((c.symm ≫ₕ f₁).restr s) ≫ₕ (f₂ ≫ₕ c') : by simp [s, trans_of_set']
+      ... ≈ ((c.symm ≫ₕ f₁) ≫ₕ (f₂ ≫ₕ c')).restr s : by simp [restr_trans]
+      ... ≈ (c.symm ≫ₕ (f₁ ≫ₕ f₂) ≫ₕ c').restr s : by simp [eq_on_source_restr, trans_assoc]
+      ... ≈ F₂ : by simp [F₂, feq],
+    have : F₂ ∈ G := G.eq_on_source F₁ F₂ A (setoid.symm this),
+    exact this
+  end,
```

Comment:
I'm slightly worried that this proof is so monolithic. Won't you repeat a lot of it in the non-invertible case (say when proving that a composition of differentiable functions is differentiable)? It looks like it could use a couple of helper lemmas, or maybe functions, or even tactics.

Reply:
This kind of computations happens repeatedly, but always in slightly different guises, so that I have not been able to extract lemmas that could be reused. Maybe tactics could help somewhat, but what kind is not clear to me. What comes up again and again is to check that a point belongs to the source of some local homeo, then its image belongs to the target of another one, and so on. This is something SMT solvers can do perfectly well, so I am not sure there is something specific to be developed here (and I am eagerly waiting for Sledgehammer :)

Reply:
I guess we need to see more of these proofs to understand what could be streamlined.